### ISSUE TITLE
Fine tuning for AIS with CLI micro_sam.train: doubts

### ISSUE THREAD

**Issue by aloboa** [computational-cell-analytics/micro-sam#939]
(continued from https://forum.image.sc/t/fine-tuning-for-ais-with-cli-micro-sam-train-doubts/110075)

Thanks. I had checked the information provided by CLI -h
but still, some points are obscure:

     --segmentation_decoder SEGMENTATION_DECODER
    Whether to finetune Segment Anything Model

But no actual options are defined. TRUE/FALSE, YES/NO, a decoder name?
Also, to make it more user-friendly, this option should rather be
--with_segmentation_decoder, in consistency with the GUI.

    --patch_shape: The patch shape used for training.

Shape of what? Which are the available options? rectangular? circle? Options are not listed. But now I use  `--patch_shape 1024 1024`

    --trained_model_name: This is just the name of the folder where your trained model will be stored.

folder? or folder + file name? Note the CLI help states: “The custom name of trained model”.

As far as I understand (see ["your model does not contain a segmentation decoder" · Issue #835 · computational-cell-analytics/micro-sam · GitHub](https://github.com/computational-cell-analytics/micro-sam/issues/835)), the trained_model_name argument is for AIS, while the --output_path is for the SAM backbone model. It would be clearer to call these arguments --AIS_trained_model_name and backbone_trained_model_name

**Comment by constantinpape** [comment]
> --segmentation_decoder SEGMENTATION_DECODER
> Whether to finetune Segment Anything Model

See the reply in #940 .

> But no actual options are defined. TRUE/FALSE, YES/NO, a decoder name?
> Also, to make it more user-friendly, this option should rather be

Thanks for the feedback. We will improve the message in `-h`.  And you can pass the name of a decoder.
Currently we only support `instances`, which is the decoder for AIS, but we will support more options in the near future.

> --with_segmentation_decoder, in consistency with the GUI.

We chose the current design to support more options, see above.

> Shape of what? Which are the available options? rectangular? circle? Options are not listed. But now I use `--patch_shape 1024 1024`

You can list two values to specify the rectangular patch shape that should be used for training. This means `1024 1024` is valid. Note that what patch shape is best depends on your problem, we usually recommend 512 x 512. We will extend the documentation.

> --trained_model_name: This is just the name of the folder where your trained model will be stored.

I think that this is the name where the model will be saved locally in `checkpoints`. We will double check this and improve the documentation.

**Comment by aloboa** [comment]
Considering both what it is said here and in https://github.com/computational-cell-analytics/micro-sam/issues/940,
the current command for AIS should either 
* omit `--segmentation_decoder` entirely (as `--segmentation_decoder` expects one argument and results in error otherwise)
* use `--segmentation_decoder instances`
`
We will improve the message in -h.`
Please do. Note that the current "Whether to finetune Segment Anything Model" seems to imply a YES/NO answer (as in the GUI).

> You can list two values to specify the rectangular patch shape that should be used for training
Then I suggest the option should be rather named `--patch_shape_dimensions `, as the shape is always rectangular. 

> we usually recommend 512 x 512
Why? Is SAM not working with 1024x1024 windows? Is 512x512 better in terms of the quality of the results or in terms of performance, or both?

Please state the default value of 512x512 in -h info.

Also:
When I start the process, I see:
```
Start fitting for 9000 iterations /  100 epochs
with 90 iterations per epoch
Training with mixed precision
```

Where is the number of iterations set?
What is "training with mixed precision"

And, for each epoch: `"Epoch 20: average [s/it]: 1.038207, current metric: 0.117544, best metric: 0.110274:"`
what is s/it? what metrics are those?

In case we had >1 image in those folders set in `--images` and `--labels`, do we need to tell micro-sam which ones are the actual pairs?
For example, if I have image1.tif, image2.tif and image3.tif in folder "images", and labels1.tif, labels2.tif and labels3.tif, in folder "labels",
how does micro-sam know that the objects labeled in labels1.tif correspond to objects in image1.tif? is there some naming convention that I am missing?
(Please not we have created our labelled images outside micro-sam)

Currentely, we have many labeled objects in 1 single relatively large image (2560 x 920). Is it better to cut into pieces so that each labeled object is in 1 single image?

Finally (by now!), does micro-sam perform any kind of training sample augmentation?

**Comment by constantinpape** [comment]
A general comment: several of your questions are related to very basic concepts of deep learning. We don't have the bandwidth to explain this in an issue or the documentation. Please always use google or ChatGPT and try to find the answer to such questions yourself before opening an issue. 50% of your questions can be easily answered this way. 

> * omit `--segmentation_decoder` entirely (as `--segmentation_decoder` expects one argument and results in error otherwise)
> 
>     * use `--segmentation_decoder instances`
> 
> 
> Correct?
> 

Yes.

> Please state the default value of 512x512 in -h info. This would solve any doubt in this option

Done in [db2ad84](https://github.com/computational-cell-analytics/micro-sam/commit/db2ad8411cc0ea680d12af81cb7936beac1b9ca0).


> * What are the default values for
> 
> 
> ```
> --model_type
> --checkpoint_path
> ```

By default, the model type define here is used: https://github.com/computational-cell-analytics/micro-sam/blob/master/micro_sam/util.py#L47 . By default we don't use a custom checkpoint path.


> * Where is the number of iterations set?

https://github.com/computational-cell-analytics/micro-sam/blob/master/micro_sam/training/training.py#L781
The number of iterations is a function of the number of epochs. Please refer to introductory courses on deep learning for the details, we cannot give a comprehensive explanation of the basis of a whole field in an issue or documentation.

> * What is "training with mixed precision"

https://pytorch.org/docs/stable/amp.html
This is a standard technique in deep learning training. Please use google  / ChatGPT before asking such questions in an issue.

> * what is s/it?

Seconds per iteration

> * what metrics are those?

A combination of dice score and the instance segmentation loss. Please refer to the methods section of our publication for details on the loss function.

> * In case we had >1 image in those folders set in `--images` and `--labels`, do we need to tell micro-sam which ones are the actual pairs?
>       For example, if I have image1.tif, image2.tif and image3.tif in folder "images", and labels1.tif, labels2.tif and labels3.tif, in folder "labels", how does micro-sam know that the objects labeled in labels1.tif correspond to objects in image1.tif? is there some naming convention that I am missing?
>       (We have created our labelled images outside micro-sam).

`micro_sam` uses a simple heuristics for this: it sorts both the images in the `image` and `labels` folder using `natsort` and then assumes that the images and labels match. So ideally you should use the same name for the files in `images` and `labels` and then they will be guaranteed to match.

> * Currently, we have many labeled objects in 1 single relatively large image (2560 x 920). Is it better to cut into pieces so that each labeled object is in 1 single image?

No, this doesn't matter, as the images will be tiled with the `patch_size`  anyways.

> * Does micro-sam perform any kind of training sample augmentation?

Yes, flips (horizontal and vertical), and rotations by 90 degree. These are applied randomly.

> * What is a labeled image in terms of format for micro-sam? any integer image?, unsigned integer 16bits? I think it is integer 32bits, either signed or unsigned, could you confirm?

Any integer datatype with 16 bits or more is valid. (8 bit is also in principle valid, but this allows at max 255 indpendent IDs, which may not be sufficient to give a unique id to each object).

> * Must background of labelled images be 0?

Yes.
